src/Commands/core/ArchiveDumpCommands.php

```diff
@@ -0,0 +1,605 @@
+<?php
+
+namespace Drush\Commands\core;
+
+use Drupal;
+use Drush\Boot\DrupalBootLevels;
+use Drush\Commands\DrushCommands;
+use Drush\Drush;
+use Drush\Sql\SqlBase;
+use Drush\Utils\FsUtils;
+use Exception;
+use FilesystemIterator;
+use Phar;
+use PharData;
+use RecursiveCallbackFilterIterator;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Process\Process;
+use Symfony\Component\Yaml\Yaml;
+use Traversable;
+use Webmozart\PathUtil\Path;
+
+/**
+ * Class ArchiveDumpCommands.
+ */
+class ArchiveDumpCommands extends DrushCommands
+{
+    /**
+     * @var \Symfony\Component\Filesystem\Filesystem
+     */
+    private Filesystem $filesystem;
+
+    /**
+     * @var string
+     */
+    private string $archiveDir;
+
+    /**
+     * @var string
+     */
+    private string $drupalFilesDir;
+
+    private const COMPONENT_CODE = 'code';
+
+    private const COMPONENT_FILES = 'files';
+
+    private const COMPONENT_DATABASE = 'database';
+    private const SQL_DUMP_FILE_NAME = 'database.sql';
+
+    private const ARCHIVES_DIR_NAME = 'archives';
+    private const ARCHIVE_FILE_NAME = 'archive.tar';
+    private const MANIFEST_FORMAT_VERSION = '1.0';
+    private const MANIFEST_FILE_NAME = 'MANIFEST.yml';
+
+    /**
+     * Backup your code, files, and database into a single file.
+     *
+     * The following paths would be excluded from a code archive:
+     *  - ".git"
+     *  - "vendor"
+     *  - "[docroot]/sites/@/settings.@.php"
+     *  - Drupal files directory
+     *  - Composer packages installed paths (`composer info --path --format=json`)
+     *
+     * The following directories would be excluded from a file archive:
+     * - css
+     * - js
+     * - styles
+     * - php
+     *
+     * @command archive:dump
+     * @validate-php-extension Phar
+     * @aliases ard
+     *
+     * @option code Archive codebase.
+     * @option files Archive Drupal files.
+     * @option db Archive database SQL dump.
+     * @option destination The full path and filename in which the archive should be stored. If omitted, it will be saved to the drush-backups directory.
+     * @option overwrite Overwrite destination file if exists.
+     * @option description Describe the archive contents.
+     * @option tags Add tags to the archive manifest. Delimit several by commas.
+     * @option overwrite Do not fail if the destination file exists; overwrite it instead. Default is --no-overwrite.
+     * @option generator The generator name to store in the MANIFEST.yml file. The default is "Drush archive-dump".
+     * @option generatorversion The generator version number to store in the MANIFEST file. The default is Drush version.
+     * @option exclude-code-paths Comma-separated list of paths to exclude from the code archive.
+     *
```

Comment:
Please add a few @usage examples. Especially --destination. Same for restore command